domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/gtfs/frequencies/Frequency.java

```diff
@@ -0,0 +1,209 @@
+         */
+        public FrequencyBuilder headwaySecs(@NotNull Integer headwaySecs) {
+            this.headwaySecs = headwaySecs;
+            return this;
+        }
+
+        /**
+         * Sets field exactTimes value and returns this
+         *
+         * @param exactTimes identifies an end time for a trip
+         * @return builder for future object creation
+         */
+        public FrequencyBuilder exactTimes(@Nullable Integer exactTimes) {
+            this.exactTimes = ExactTimes.fromInt(exactTimes);
+            this.originalExactTimes = exactTimes;
+            return this;
+        }
+
+        public EntityBuildResult<?> build() {
+            noticeCollection.clear();
+            if (tripId == null || startTime == null || endTime == null || headwaySecs == null ||
+                    headwaySecs < 0 || exactTimes == null) {
+
+                if (tripId == null) {
+                    noticeCollection.add(new MissingRequiredValueNotice("frequencies.txt",
+                            "trip_id", tripId));
+                }
+
+                if (startTime == null) {
+                    noticeCollection.add(new MissingRequiredValueNotice("frequencies.txt",
+                            "start_time", tripId));
+                }
+                //else if (startTime == null && originalStartTime != null) InvalidTimeNotice?
+
+                if (endTime == null) {
+                    noticeCollection.add(new MissingRequiredValueNotice("frequencies.txt",
+                            "end_time", tripId));
+                }
+                //else if (endTime == null && originalEndTime != null) InvalidTimeNotice?
+
+                if (headwaySecs == null) {
+                    noticeCollection.add(new MissingRequiredValueNotice("frequencies.txt",
+                            "headway_secs", tripId));
+                } else if (headwaySecs < 0) {
+                    noticeCollection.add(new IntegerFieldValueOutOfRangeNotice("frequencies.txt",
+                            "headway_secs", tripId, 0, Integer.MAX_VALUE, headwaySecs));
+                }
+
+                // the following statement is true when ExactTimes.isEnumValueValid(originalTransferInteger)
+                // returns false
+                if (exactTimes == null) {
+                    noticeCollection.add(new UnexpectedEnumValueNotice("frequencies.txt",
+                            "exact_times", tripId, originalExactTimes));
+                }
+                return new EntityBuildResult<>(noticeCollection);
+            }
+            return new EntityBuildResult<>(new Frequency(tripId, startTime, endTime, headwaySecs, exactTimes));
+        }
+    }
+
+    /**
+     * Returns the key corresponding to this {@link FareRule}
+     *
+     * @return the key corresponding to this {@link FareRule}
+     */
+    public static String getFrequencyMappingKey(final String tripId, final Integer startTime) {
```

Comment:
If we map `Frequency` entities on `trip_id`+`start_time` shouldn't we use this same key as `entity_id` when generating notices?

Reply:
Good point, should we create a human readable string as we pass the entity id to the notice ex. `"" + tripId + ", " + startTime`? Or create a method?

Reply:
Also, this is not implemented for other entities with similar case (ex. FareRule). 

@barbeau @fabrice-v What do you think? Should we go forward with this for every entity?

Maybe a `getHumanReadableUniqueKey()` method for every entity could be interesting if it is to be recurrent.

Reply:
@lionel-nj Yes, if we're identifying entities uniquely internally with a combination of fields, then those same fields should be shared in any notices to let the user uniquely identify where the error is coming from in the GTFS file.

@maximearmstrong In the RT validator we ended with similar helper methods to generate the "field name + value" as human-readable text in errors. So the output in the notice would include the text (in some form) "...in frequencies.txt trip_id 100 startTime 12:12:12" to point the user to this record in the file.